storage: tocommit is out of range

backup2tb crashed in this nightly run: https://teamcity.cockroachdb.com/viewLog.html?buildId=586079&buildTypeId=Cockroach_Nightlies_WorkloadNightly&tab=buildLog

At 11:40, we see a very concerning raft problem on n10 that brings the node down.

```
E180329 11:40:01.157113 263 util/log/crash_reporting.go:451  [n1,s1] Reported as error a24b491ae75e4888b3ee0f3b35da4a1b
panic: tocommit(26) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost? [recovered]
	panic: tocommit(26) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost?

goroutine 263 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc42002a240, 0x26120a0, 0xc427505470)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:177 +0x11f
panic(0x1fddd20, 0xc426b888d0)
	/usr/local/go/src/runtime/panic.go:505 +0x229
github.com/cockroachdb/cockroach/pkg/storage.(*raftLogger).Panicf(0xc4267fe900, 0x23a90b0, 0x5d, 0xc44b994260, 0x2, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/raft.go:106 +0xfe
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*raftLog).commitTo(0xc430c9b110, 0x1a)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/log.go:191 +0x15c
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*raft).handleHeartbeat(0xc42e1a4400, 0x8, 0x4, 0x1, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/raft.go:1232 +0x54
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.stepFollower(0xc42e1a4400, 0x8, 0x4, 0x1, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/raft.go:1177 +0x44c
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*raft).Step(0xc42e1a4400, 0x8, 0x4, 0x1, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/raft.go:898 +0x1308
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*RawNode).Step(0xc442924fc0, 0x8, 0x4, 0x1, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/rawnode.go:195 +0xdd
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).stepRaftGroup.func1(0xc442924fc0, 0x2612020, 0xc4200400c0, 0x26120a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:3329 +0xa8
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).withRaftGroupLocked(0xc44ac34000, 0x2418800, 0xc42c4e1b88, 0xc42c4e1b70, 0x6fe709)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:607 +0x10f
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).withRaftGroup(0xc44ac34000, 0xc42c4e1b88, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:624 +0x88
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).stepRaftGroup(0xc44ac34000, 0xc4568b36c0, 0x2099860, 0x30)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:3324 +0x50
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRaftRequestWithReplica(0xc420819000, 0x26120a0, 0xc43dbf5a10, 0xc44ac34000, 0xc4568b36c0, 0xc43dbf5a10)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3175 +0xeb
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRequestQueue.func1(0x26120a0, 0xc43dbf5a10, 0xc44ac34000, 0x26120a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3793 +0x97
github.com/cockroachdb/cockroach/pkg/storage.(*Store).withReplicaForRequest(0xc420819000, 0x26120a0, 0xc43dbf5a10, 0xc4568b36c0, 0xc42c4e1ed0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3121 +0x135
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRequestQueue(0xc420819000, 0x26120a0, 0xc427505470, 0x706c)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3787 +0x229
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc420b80000, 0x26120a0, 0xc427505470)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:226 +0x21b
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2(0x26120a0, 0xc427505470)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:166 +0x3e
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc4262f8ec0, 0xc42002a240, 0xc4262f8eb0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:192 +0xe9
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:185 +0xad
```

https://teamcity.cockroachdb.com/repository/download/Cockroach_Nightlies_WorkloadNightly/586079:id/20180328-586079/backup2TB/10.logs/cockroach.log

Two minutes later, the SSH connection to n1, the node coordinating the backup, gets OOM killed (!).

```
[11:42:11]	[Step 3/3] 	cluster.go:642,backup.go:52,cluster.go:784,errgroup.go:58: roachprod ssh teamcity-586079-backup2tb:1 -- ./cockroach sql --insecure -e "
[11:42:11]	[Step 3/3] 						BACKUP bank.bank TO 'gs://cockroachdb-backup-testing/teamcity-586079-backup2tb'": signal: killed
```

There's not a whole lot in n1's logs:

```
I180329 11:40:01.353808 1350 storage/store.go:3470  [n1,s1,r349/5:/Table/51/1/470{0456/0-3729}] added to replica GC queue (contacted deleted peer)
W180329 11:40:01.565404 308031 vendor/google.golang.org/grpc/server.go:961  grpc: Server.processUnaryRPC failed to write status connection error: desc = "transport is closing"
W180329 11:40:01.568787 429 vendor/google.golang.org/grpc/clientconn.go:1158  grpc: addrConn.createTransport failed to connect to {teamcity-586079-backup2tb-0010:26257 0  <nil>}. Err :connection error: desc = "transport: Error while dialing cannot reuse client connection". Reconnecting...
W180329 11:40:01.568830 429 vendor/google.golang.org/grpc/clientconn.go:1277  grpc: addrConn.transportMonitor exits due to: context canceled
W180329 11:40:01.674317 848092 vendor/google.golang.org/grpc/server.go:625  grpc: Server.Serve failed to create ServerTransport: connection error: desc = "transport: http2Server.HandleStreams failed to receive the preface from client: EOF"
```

https://teamcity.cockroachdb.com/repository/download/Cockroach_Nightlies_WorkloadNightly/586079:id/20180328-586079/backup2TB/1.logs/cockroach.log

/cc @vilterp

--- a/pkg/ccl/workloadccl/fixture.go
+++ b/pkg/ccl/workloadccl/fixture.go
@@ -107,6 +107,11 @@ func generatorToGCSFolder(config FixtureConfig, gen workload.Generator) string {
 	)
 }
 
+// FixtureURL returns the URL for pre-computed Generator data stored on GCS.
+func FixtureURL(config FixtureConfig, gen workload.Generator) string {
+	return fmt.Sprintf("gs://%s/%s", config.GCSBucket, generatorToGCSFolder(config, gen))
+}
+
 // GetFixture returns a handle for pre-computed Generator data stored on GCS. It
 // is expected that the generator will have had Configure called on it.
 func GetFixture(

--- a/pkg/cmd/roachtest/main.go
+++ b/pkg/cmd/roachtest/main.go
@@ -30,26 +30,38 @@ func main() {
 		Short: "roachtest tool for testing cockroach clusters",
 		Long: `roachtest is a tool for testing cockroach clusters.
 `,
+
+		PersistentPreRunE: func(*cobra.Command, []string) error {
+			if clusterName != "" && local {
+				return fmt.Errorf("cannot specify both an existing cluster (%s) and --local", clusterName)
+			}
+			initBinaries()
+			return nil
+		},
 	}
+
+	rootCmd.PersistentFlags().StringVarP(
+		&clusterName, "cluster", "c", "", "name of an existing cluster to use for running tests")
+	rootCmd.PersistentFlags().BoolVarP(
+		&local, "local", "l", local, "run tests locally")
+	rootCmd.PersistentFlags().StringVarP(
+		&username, "user", "u", username, "username to run under, detect if blank")
+	rootCmd.PersistentFlags().StringVar(
+		&cockroach, "cockroach", "", "path to cockroach binary to use")
+	rootCmd.PersistentFlags().StringVar(
+		&workload, "workload", "", "path to workload binary to use")
 
 	var runCmd = &cobra.Command{
 		Use:   "run [tests]",
-		Short: "run automated tests on cockroach cluster\n",
+		Short: "run automated tests on cockroach cluster",
 		Long: `Run automated tests on existing or ephemeral cockroach clusters.
 
 Use 'roachtest run -n' to see a list of all tests.
 `,
 		RunE: func(_ *cobra.Command, args []string) error {
-			if clusterName != "" && local {
-				return fmt.Errorf("cannot specify both an existing cluster (%s) and --local", clusterName)
-			}
-
 			if count <= 0 {
 				return fmt.Errorf("--count (%d) must by greater than 0", count)
 			}
-
-			initBinaries()
-
 			r := newRegistry()
 			registerTests(r)
 			os.Exit(r.Run(args))
@@ -65,31 +77,42 @@ Use 'roachtest run -n' to see a list of all tests.
 		&debug, "debug", "d", debug, "don't wipe and destroy cluster if test fails")
 	runCmd.Flags().BoolVarP(
 		&dryrun, "dry-run", "n", dryrun, "dry run (don't run tests)")
-	runCmd.Flags().BoolVarP(
-		&local, "local", "l", local, "run tests locally")
 	runCmd.Flags().IntVarP(
 		&parallelism, "parallelism", "p", parallelism, "number of tests to run in parallel")
 	runCmd.Flags().StringVar(
 		&artifacts, "artifacts", "artifacts", "path to artifacts directory")
-	runCmd.Flags().StringVarP(
-		&clusterName, "cluster", "c", "", "name of an existing cluster to use for running tests")
 	runCmd.Flags().StringVar(
 		&clusterID, "cluster-id", "", "an identifier to use in the test cluster's name")
 	runCmd.Flags().BoolVar(
 		&clusterWipe, "wipe", true,
 		"wipe existing cluster before starting test (for use with --cluster)")
 	runCmd.Flags().StringVar(
-		&cockroach, "cockroach", "", "path to cockroach binary to use")
-	runCmd.Flags().StringVarP(
-		&username, "user", "u", username, "username to run under, detect if blank")
-	runCmd.Flags().StringVar(
-		&workload, "workload", "", "path to workload binary to use")
-	runCmd.Flags().StringVar(
 		&slackToken, "slack-token", "", "Slack bot token")
 	runCmd.Flags().StringVar(
 		&zones, "zones", "", "Zones for the cluster (use roachprod defaults if empty)")
 	runCmd.Flags().BoolVar(
 		&teamCity, "teamcity", false, "include teamcity-specific markers in output")
+
+	var storeGenCmd = &cobra.Command{
+		Use:   "store-gen [workload]",
+		Short: "generate store directory dumps\n",
+		Long: `Generate store directory dumps that can quickly bootstrap a
+	Cockroach cluster with existing data.
+	`,
+		Args: cobra.MinimumNArgs(1),
+		RunE: func(_ *cobra.Command, args []string) error {
+			r := newRegistry()
+			registerStoreGen(r, args)
+			// We've only registered one store generation "test" that does its own
+			// argument processing, so no need to provide any arguments to r.Run.
+			os.Exit(r.Run(nil /* filter */))
+			return nil
+		},
+	}
+	storeGenCmd.Flags().IntVarP(
+		&stores, "stores", "n", stores, "number of stores to distribute data across")
+	storeGenCmd.Flags().SetInterspersed(false) // ignore workload flags
+	rootCmd.AddCommand(storeGenCmd)
 
 	if err := rootCmd.Execute(); err != nil {
 		// Cobra has already printed the error message.

--- a/pkg/cmd/roachtest/store_gen.go
+++ b/pkg/cmd/roachtest/store_gen.go
@@ -1,0 +1,87 @@
+// Copyright 2018 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+
+	"golang.org/x/sync/errgroup"
+)
+
+var stores = 3
+
+// registerStoreGen registers a store generation "test" that powers the
+// 'roachtest store-gen' subcommand.
+func registerStoreGen(r *registry, args []string) {
+	workloadArgs := strings.Join(args, " ")
+
+	r.Add(testSpec{
+		Name:  "store-gen",
+		Nodes: nodes(stores),
+		Run: func(ctx context.Context, t *test, c *cluster) {
+			c.Put(ctx, cockroach, "./cockroach")
+			c.Put(ctx, workload, "./workload")
+			c.Start(ctx)
+
+			m := newMonitor(ctx, c)
+			m.Go(func(ctx context.Context) error {
+				t.Status("restoring fixture")
+				c.Run(ctx, c.Node(1), fmt.Sprintf("./workload fixtures load %s", workloadArgs))
+
+				urlBase, err := c.RunWithBuffer(ctx, c.l, c.Node(1),
+					fmt.Sprintf(`./workload fixtures url %s`, workloadArgs))
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				db := c.Conn(ctx, 1)
+				defer db.Close()
+				var binVersion string
+				err = db.QueryRow("SELECT crdb_internal.node_executable_version()").Scan(&binVersion)
+				if err != nil {
+					return err
+				}
+
+				fixturesURL := fmt.Sprintf("%s/stores=%d,bin-version=%s",
+					bytes.TrimSpace(urlBase), stores, binVersion)
+
+				// gsutil rm fails if no objects exist, even with -f, so check if any
+				// old store directories exist before attempting to remove them.
+				err = c.RunE(ctx, c.Node(1), fmt.Sprintf("gsutil -m -q ls %s &> /dev/null", fixturesURL))
+				if err == nil {
+					t.Status("deleting old store dumps")
+					c.Run(ctx, c.Node(1), fmt.Sprintf("gsutil -m -q rm -r %s", fixturesURL))
+				}
+
+				t.Status("uploading store dumps")
+				var g errgroup.Group
+				for store := 1; store <= stores; store++ {
+					store := store
+					g.Go(func() error {
+						c.Run(ctx, c.Node(store), fmt.Sprintf("gsutil -m -q cp -r {store-dir}/* %s/%d/",
+							fixturesURL, store))
+						return nil
+					})
+				}
+				return g.Wait()
+			})
+			m.Wait()
+		},
+	})
+}

--- a/pkg/cmd/workload/check.go
+++ b/pkg/cmd/workload/check.go
@@ -27,10 +27,10 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/workload"
 )
 
-var checkCmd = &cobra.Command{
+var checkCmd = setCmdDefaults(&cobra.Command{
 	Use:   `check`,
 	Short: `Check a running cluster's data for consistency`,
-}
+})
 
 func init() {
 	for _, meta := range workload.Registered() {
@@ -53,10 +53,10 @@ func init() {
 			}
 		}
 
-		genCheckCmd := &cobra.Command{
+		genCheckCmd := setCmdDefaults(&cobra.Command{
 			Use:  meta.Name + ` [CRDB URI]`,
 			Args: cobra.RangeArgs(0, 1),
-		}
+		})
 		genCheckCmd.Flags().AddFlagSet(genFlags)
 		genCheckCmd.Run = cmdHelper(gen, check)
 		checkCmd.AddCommand(genCheckCmd)

--- a/pkg/cmd/workload/csv_server.go
+++ b/pkg/cmd/workload/csv_server.go
@@ -25,12 +25,12 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/workload"
 )
 
-var csvServerCmd = &cobra.Command{
+var csvServerCmd = setCmdDefaults(&cobra.Command{
 	Use:   `csv-server`,
 	Short: `Serves csv table data through an HTTP interface`,
 	Args:  cobra.NoArgs,
 	RunE:  runCSVServer,
-}
+})
 
 var port *int
 

--- a/pkg/cmd/workload/fixtures.go
+++ b/pkg/cmd/workload/fixtures.go
@@ -67,6 +67,10 @@ var fixturesLoadCmd = setCmdDefaults(&cobra.Command{
 	Use:   `load`,
 	Short: `Load a fixture into a running cluster. An enterprise license is required.`,
 })
+var fixturesURLCmd = setCmdDefaults(&cobra.Command{
+	Use:   `url`,
+	Short: `Generate the GCS URL for a fixture`,
+})
 
 var fixturesMakeCSVServerURL = fixturesMakeCmd.PersistentFlags().String(
 	`csv-server`, ``,
@@ -135,10 +139,19 @@ func init() {
 		genLoadCmd.Flags().AddFlagSet(genFlags)
 		genLoadCmd.Run = cmdHelper(gen, fixturesLoad)
 		fixturesLoadCmd.AddCommand(genLoadCmd)
+
+		genURLCmd := setCmdDefaults(&cobra.Command{
+			Use:  meta.Name,
+			Args: cobra.NoArgs,
+		})
+		genURLCmd.Flags().AddFlagSet(genFlags)
+		genURLCmd.Run = fixturesURL(gen)
+		fixturesURLCmd.AddCommand(genURLCmd)
 	}
 	fixturesCmd.AddCommand(fixturesListCmd)
 	fixturesCmd.AddCommand(fixturesMakeCmd)
 	fixturesCmd.AddCommand(fixturesLoadCmd)
+	fixturesCmd.AddCommand(fixturesURLCmd)
 	rootCmd.AddCommand(fixturesCmd)
 }
 
@@ -248,4 +261,17 @@ func fixturesLoad(gen workload.Generator, urls []string, dbName string) error {
 	}
 
 	return nil
+}
+
+func fixturesURL(gen workload.Generator) func(*cobra.Command, []string) {
+	return handleErrs(func(*cobra.Command, []string) error {
+		if h, ok := gen.(workload.Hookser); ok {
+			if err := h.Hooks().Validate(); err != nil {
+				return err
+			}
+		}
+
+		fmt.Println(workloadccl.FixtureURL(config(), gen))
+		return nil
+	})
 }

--- a/pkg/cmd/workload/fixtures.go
+++ b/pkg/cmd/workload/fixtures.go
@@ -51,20 +51,22 @@ func config() workloadccl.FixtureConfig {
 	return config
 }
 
-var fixturesCmd = &cobra.Command{Use: `fixtures`}
-var fixturesListCmd = &cobra.Command{
+var fixturesCmd = setCmdDefaults(&cobra.Command{
+	Use: `fixtures`,
+})
+var fixturesListCmd = setCmdDefaults(&cobra.Command{
 	Use:   `list`,
 	Short: `List all fixtures stored on GCS`,
 	Run:   handleErrs(fixturesList),
-}
-var fixturesMakeCmd = &cobra.Command{
+})
+var fixturesMakeCmd = setCmdDefaults(&cobra.Command{
 	Use:   `make`,
 	Short: `Regenerate and store a fixture on GCS`,
-}
-var fixturesLoadCmd = &cobra.Command{
+})
+var fixturesLoadCmd = setCmdDefaults(&cobra.Command{
 	Use:   `load`,
 	Short: `Load a fixture into a running cluster. An enterprise license is required.`,
-}
+})
 
 var fixturesMakeCSVServerURL = fixturesMakeCmd.PersistentFlags().String(
 	`csv-server`, ``,
@@ -118,18 +120,18 @@ func init() {
 			}
 		}
 
-		genMakeCmd := &cobra.Command{
+		genMakeCmd := setCmdDefaults(&cobra.Command{
 			Use:  meta.Name + ` [CRDB URI]`,
 			Args: cobra.RangeArgs(0, 1),
-		}
+		})
 		genMakeCmd.Flags().AddFlagSet(genFlags)
 		genMakeCmd.Run = cmdHelper(gen, fixturesMake)
 		fixturesMakeCmd.AddCommand(genMakeCmd)
 
-		genLoadCmd := &cobra.Command{
+		genLoadCmd := setCmdDefaults(&cobra.Command{
 			Use:  meta.Name + ` [CRDB URI]`,
 			Args: cobra.RangeArgs(0, 1),
-		}
+		})
 		genLoadCmd.Flags().AddFlagSet(genFlags)
 		genLoadCmd.Run = cmdHelper(gen, fixturesLoad)
 		fixturesLoadCmd.AddCommand(genLoadCmd)

--- a/pkg/cmd/workload/main.go
+++ b/pkg/cmd/workload/main.go
@@ -23,9 +23,9 @@ import (
 	_ "github.com/cockroachdb/cockroach/pkg/ccl/workloadccl/allccl"
 )
 
-var rootCmd = &cobra.Command{
+var rootCmd = setCmdDefaults(&cobra.Command{
 	Use: `workload`,
-}
+})
 
 func main() {
 	if err := rootCmd.Execute(); err != nil {

--- a/pkg/cmd/workload/run.go
+++ b/pkg/cmd/workload/run.go
@@ -40,10 +40,10 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/workload"
 )
 
-var runCmd = &cobra.Command{
+var runCmd = setCmdDefaults(&cobra.Command{
 	Use:   `run`,
 	Short: `Run a workload's operations against a cluster`,
-}
+})
 
 var runFlags = pflag.NewFlagSet(`run`, pflag.ContinueOnError)
 var tolerateErrors = runFlags.Bool("tolerate-errors", false, "Keep running on error")
@@ -54,10 +54,10 @@ var duration = runFlags.Duration("duration", 0, "The duration to run. If 0, run 
 var doInit = runFlags.Bool("init", false, "Automatically run init")
 var ramp = runFlags.Duration("ramp", 0*time.Second, "The duration over which to ramp up load.")
 
-var initCmd = &cobra.Command{
+var initCmd = setCmdDefaults(&cobra.Command{
 	Use:   `init`,
 	Short: `Set up tables for a workload`,
-}
+})
 
 var initFlags = pflag.NewFlagSet(`init`, pflag.ContinueOnError)
 var drop = initFlags.Bool("drop", false, "Drop the existing database, if it exists")
@@ -76,13 +76,13 @@ func init() {
 			genFlags = f.Flags().FlagSet
 		}
 
-		genInitCmd := &cobra.Command{Use: meta.Name, Short: meta.Description}
+		genInitCmd := setCmdDefaults(&cobra.Command{Use: meta.Name, Short: meta.Description})
 		genInitCmd.Flags().AddFlagSet(initFlags)
 		genInitCmd.Flags().AddFlagSet(genFlags)
 		genInitCmd.Run = cmdHelper(gen, runInit)
 		initCmd.AddCommand(genInitCmd)
 
-		genRunCmd := &cobra.Command{Use: meta.Name, Short: meta.Description}
+		genRunCmd := setCmdDefaults(&cobra.Command{Use: meta.Name, Short: meta.Description})
 		genRunCmd.Flags().AddFlagSet(runFlags)
 		genRunCmd.Flags().AddFlagSet(genFlags)
 		initFlags.VisitAll(func(initFlag *pflag.Flag) {
@@ -126,6 +126,23 @@ func cmdHelper(
 		}
 		return fn(gen, urls, dbName)
 	})
+}
+
+// setCmdDefaults ensures that the provided Cobra command will properly report
+// an error if the user specifies an invalid subcommand. It is safe to call on
+// any Cobra command.
+//
+// This is a wontfix bug in Cobra: https://github.com/spf13/cobra/pull/329
+func setCmdDefaults(cmd *cobra.Command) *cobra.Command {
+	if cmd.Run == nil && cmd.RunE == nil {
+		cmd.Run = func(cmd *cobra.Command, args []string) {
+			_ = cmd.Usage()
+		}
+	}
+	if cmd.Args == nil {
+		cmd.Args = cobra.NoArgs
+	}
+	return cmd
 }
 
 // numOps keeps a global count of successful operations.

